mobile/test/java/org/chromium/net/CronetHttp3Test.java

```diff
@@ -0,0 +1,151 @@
+import org.chromium.net.testing.Feature;
+import org.chromium.net.testing.TestUrlRequestCallback;
+import org.chromium.net.testing.TestUrlRequestCallback.ResponseStep;
+import io.envoyproxy.envoymobile.engine.JniLibrary;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import io.envoyproxy.envoymobile.engine.testing.HttpTestServerFactory;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Collections;
+import java.io.File;
+import java.nio.file.Paths;
+import java.nio.file.Files;
+
+/**
+ * Test CronetEngine with production HTTP/3 logic
+ */
+@RunWith(RobolectricTestRunner.class)
+public class CronetHttp3Test {
+  @Rule public final CronetTestRule mTestRule = new CronetTestRule();
+
+  private static final String TAG = CronetHttp3Test.class.getSimpleName();
+
+  // URLs used for tests.
+
+  // If true, dump envoy logs on test completion.
+  // Ideally we could override this from the command line but that's TBD.
+  boolean mPrintEnvoyLogs = false;
+  // The HTTP/2 server, set up to alt-svc to the HTTP/3 server
+  private HttpTestServerFactory.HttpTestServer mHttp2TestServer;
+  // The HTTP/3 server
+  private HttpTestServerFactory.HttpTestServer mHttp3TestServer;
+  // An optional CronvoyLogger, set up if mPrintEnvoyLogs is true.
+  private CronvoyLogger mLogger;
+  // The log file name,  set up if mPrintEnvoyLogs is true.
+  private String mLogFilename;
+  // The engine for this test.
+  CronvoyUrlRequestContext mCronvoyEngine;
+
+  @BeforeClass
+  public static void loadJniLibrary() {
+    JniLibrary.loadTestLibrary();
+  }
+
+  public void setUp(boolean setUpLogging) throws Exception {
+    // Set up the HTTP/3 server
+    Map<String, String> headers = new HashMap<>();
+    mHttp3TestServer = HttpTestServerFactory.start(HttpTestServerFactory.Type.HTTP3, 0, headers,
+                                                   "This is a simple text file served by QUIC.\n",
+                                                   Collections.emptyMap());
+    // Next set up the HTTP/2 server, advertising HTTP/3 support for the HTTP/3 server
+    String alt_svc = "h3=\":" + mHttp3TestServer.getPort() + "\"; ma=86400";
+    headers.put("alt-svc", alt_svc);
+    // Note that the HTTP/2 server must start on the same port as Envoy currently does not accept
+    // alt-svc with differing ports. This may cause problems if this UDP port is in use at which
+    // point listening on 127.0.0.N where N!=1 may improve flakiness.
+    mHttp2TestServer = HttpTestServerFactory.start(
+        HttpTestServerFactory.Type.HTTP2_WITH_TLS, mHttp3TestServer.getPort(), headers,
+        "This is a simple text file served by QUIC.\n", Collections.emptyMap());
+
+    // Optionally, set up logging. This will slow down the tests a bit but make debugging much
+    // easier.
+    if (setUpLogging) {
```

Comment:
Instead of using netlog, there's another way to show Envoy logs via

```
private CronvoyLogger mLogger  = new CronvoyLogger() {
  @Override
  public void log(int logLevel, String message) {
    System.out.print(message);
  }
}
```

Reply:
very good call - done